File: apps/python/lib/chapter01/kinoko_takenoko.py
import csv
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class Person:
    height: int
    weight: int
    age_group: int
    faction: str


@dataclass(frozen=True)
class Features:
    height: int
    weight: int
    age_group: int


def load_people(csv_file: Path) -> list[Person]:
    with csv_file.open(encoding="utf-8-sig", newline="") as f:
        return [
            Person(
                height=int(row["身長"]),
                weight=int(row["体重"]),
                age_group=int(row["年代"]),
                faction=row["派閥"],
            )
            for row in csv.DictReader(f)
        ]


def split_features_and_labels(people: list[Person]) -> tuple[list[Features], list[str]]:
    features = [Features(p.height, p.weight, p.age_group) for p in people]
    labels = [p.faction for p in people]
    return features, labels


def predict_by_rule(features: Features) -> str:
    if features.age_group == 20:
        return "きのこ"
    return "たけのこ"


def accuracy(predictions: list[str], labels: list[str]) -> float:
    correct = sum(1 for p, t in zip(predictions, labels, strict=True) if p == t)
    return correct / len(labels)

File: apps/java/src/main/java/chapter01/KinokoTakenoko.java
package chapter01;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/** きのこ派・たけのこ派の判定。 */
public final class KinokoTakenoko {
  private static final String BOM = "﻿";

  /** 「20 代ならきのこ派」というルールの年代 */
  private static final int KINOKO_AGE_GROUP = 20;

  private KinokoTakenoko() {}

  /** BOM 付きの UTF-8 の CSV を読み込み、列名で値を取り出して人物のリストにする。 */
  public static List<Person> loadPeople(Path csvFile) throws IOException {
    List<String> lines = Files.readAllLines(csvFile, StandardCharsets.UTF_8);
    List<String> header = Arrays.asList(stripBom(lines.getFirst()).split(","));
    Map<String, Integer> index =
        IntStream.range(0, header.size())
            .boxed()
            .collect(Collectors.toMap(header::get, Function.identity()));
    return lines.stream()
        .skip(1)
        .filter(line -> !line.isBlank())
        .map(line -> line.split(","))
        .map(
            values ->
                new Person(
                    Integer.parseInt(values[index.get("身長")]),
                    Integer.parseInt(values[index.get("体重")]),
                    Integer.parseInt(values[index.get("年代")]),
                    values[index.get("派閥")]))
        .toList();
  }

  private static String stripBom(String line) {
    return line.startsWith(BOM) ? line.substring(BOM.length()) : line;
  }

  /** 人物のリストを特徴量と正解ラベルに分ける。 */
  public static FeaturesAndLabels splitFeaturesAndLabels(List<Person> people) {
    List<Features> features =
        people.stream().map(p -> new Features(p.height(), p.weight(), p.ageGroup())).toList();
    List<String> labels = people.stream().map(Person::faction).toList();
    return new FeaturesAndLabels(features, labels);
  }

  /** 人間が決めたルールで派閥を判定する。 */
  public static String predictByRule(Features features) {
    return features.ageGroup() == KINOKO_AGE_GROUP ? "きのこ" : "たけのこ";
  }

  /** 予測が正解ラベルと一致した割合を返す。 */
  public static double accuracy(List<String> predictions, List<String> labels) {
    if (predictions.size() != labels.size()) {
      throw new IllegalArgumentException("予測と正解ラベルの件数が違います");
    }
    long correct =
        IntStream.range(0, labels.size())
            .filter(i -> predictions.get(i).equals(labels.get(i)))
            .count();
    return (double) correct / labels.size();
  }
}
